Improved error checking

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35058 C++ smoketests for the API.
* #34668 Python-driven testing for the new operator registration API
* #34620 Minor OperatorName cleanup
* **#34618 Improved error checking**
* #34601 Implement schema matching inside Dispatcher
* #34456 Finish implementing schema-less impl, switch all internal registrations to it
* #34455 Add torch::jit::parseName (mirroring parseSchema/parseSchemaOrName)
* #34221 Relax internal invariants of Dispatcher in preparation for schema-less registration

- Bring back alias analysis kind check when you def() an operator multiple times.
  It is an error to rely on alias analysis kind merging.  Fixed up op_registration_test
  to respect this.
- Add new syntax sugar dispatch_autograd for registering autograd kernels
- Don't generate schema registration for manual registrations
  (function_wrapper.py)
- Switch VariableType registrations to new API
- Switch mobile autograd to new API

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396128](https://our.internmc.facebook.com/intern/diff/D20396128)

ezyang committed:
Improved error checking

- Bring back alias analysis kind check when you def() an operator multiple times.
  It is an error to rely on alias analysis kind merging.  Fixed up op_registration_test
  to respect this.
- Add new syntax sugar dispatch_autograd for registering autograd kernels
- Don't generate schema registration for manual registrations
  (function_wrapper.py)
- Switch VariableType registrations to new API
- Switch mobile autograd to new API

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
cc @ailzhang @dlibenzi this changeset will require XLA to switch to the new style registration API. Not ready yet to prepare a PR for XLA, but will need one soon.

ezyang committed:
Update on "Improved error checking"


- Bring back alias analysis kind check when you def() an operator multiple times.
  It is an error to rely on alias analysis kind merging.  Fixed up op_registration_test
  to respect this.
- Add new syntax sugar dispatch_autograd for registering autograd kernels
- Don't generate schema registration for manual registrations
  (function_wrapper.py)
- Switch VariableType registrations to new API
- Switch mobile autograd to new API

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396128](https://our.internmc.facebook.com/intern/diff/D20396128)

[ghstack-poisoned]

ezyang committed:
Update on "Improved error checking"


- Bring back alias analysis kind check when you def() an operator multiple times.
  It is an error to rely on alias analysis kind merging.  Fixed up op_registration_test
  to respect this.
- Add new syntax sugar dispatch_autograd for registering autograd kernels
- Don't generate schema registration for manual registrations
  (function_wrapper.py)
- Switch VariableType registrations to new API
- Switch mobile autograd to new API

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396128](https://our.internmc.facebook.com/intern/diff/D20396128)

[ghstack-poisoned]

ezyang committed:
Update on "Improved error checking"


- Bring back alias analysis kind check when you def() an operator multiple times.
  It is an error to rely on alias analysis kind merging.  Fixed up op_registration_test
  to respect this.
- Add new syntax sugar dispatch_autograd for registering autograd kernels
- Don't generate schema registration for manual registrations
  (function_wrapper.py)
- Switch VariableType registrations to new API
- Switch mobile autograd to new API

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396128](https://our.internmc.facebook.com/intern/diff/D20396128)

[ghstack-poisoned]

ezyang committed:
Update on "Improved error checking"


- Bring back alias analysis kind check when you def() an operator multiple times.
  It is an error to rely on alias analysis kind merging.  Fixed up op_registration_test
  to respect this.
- Add new syntax sugar dispatch_autograd for registering autograd kernels
- Don't generate schema registration for manual registrations
  (function_wrapper.py)
- Switch VariableType registrations to new API
- Switch mobile autograd to new API

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396128](https://our.internmc.facebook.com/intern/diff/D20396128)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -102,6 +102,33 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
         op.operatorIterator_->op.registerSchema(std::move(schema));
       } else {
         TORCH_CHECK(op.schema() == schema, "Tried to register multiple operators with the same name and the same overload name but different schemas: ", schema, " vs ", op.schema());
+        // If the *new* schema is the default alias analysis kind, for BC, we
```
Dispatcher::registerDef is becoming very long, can we split this into named subfunctions for readability?

ezyang replied:
This feels a little unfair: most of the lines are devoted to making good error messages :) Do you have an alternative structure in mind?

smessmer commented on aten/src/ATen/core/dispatch/Dispatcher.cpp:
```diff
@@ -102,6 +102,33 @@ RegistrationHandleRAII Dispatcher::registerDef(FunctionSchema schema) {
+        // means that you may nondeterministically fail the subsequent test.
+        if (schema.isDefaultAliasAnalysisKind()) {
+          // OK
+        // If you POST-FACTO specify a non-default alias analysis kind after
```
nit: indentation is weird. Maybe add the comments into the then/else block instead of before the block

smessmer commented on aten/src/ATen/core/op_registration/op_registration_test.cpp:
```diff
@@ -42,19 +42,14 @@ struct MockKernel final : OperatorKernel {
 TEST(OperatorRegistrationTest, whenRegisteringSameSchemaWithAliasAnalysisAfterRegisteringWithoutAliasAnalysis_thenCanBeCalled) {
   {
     auto registrar1 = c10::RegisterOperators().op("_test::dummy(Tensor dummy) -> ()", c10::RegisterOperators::options().kernel<DummyKernel>(c10::DispatchKey::CPUTensorId));
-    auto registrar2 = c10::RegisterOperators().op("_test::dummy(Tensor dummy) -> ()", c10::RegisterOperators::options().kernel<DummyKernel>(c10::DispatchKey::XLATensorId).aliasAnalysis(at::AliasAnalysisKind::PURE_FUNCTION));
```
I think this is not doing what the test name says anymore.

ezyang replied:
I think I just need to rename the test name here

ezyang replied:
split this into two tests and lined up the names

smessmer commented on test/test_cpp_extensions_jit.py:
```diff
@@ -365,19 +365,6 @@ def test_lenient_flag_handling_in_jit_extensions(self):
         z = module.tanh_add(x, y).cpu()
         self.assertEqual(z, x.tanh() + y.tanh())
 
-    def test_complex_registration(self):
```
What was the original reason for this test? Are we sure we don't need it anymore?

ezyang replied:
IT was originally to test that the complex out-of-tree extension point was working. But since Anjali is working on putting complex in tree we don't need this extension point anymore.

ezyang committed:
Update on "Improved error checking"


- Bring back alias analysis kind check when you def() an operator multiple times.
  It is an error to rely on alias analysis kind merging.  Fixed up op_registration_test
  to respect this.
- Add new syntax sugar dispatch_autograd for registering autograd kernels
- Don't generate schema registration for manual registrations
  (function_wrapper.py)
- Switch VariableType registrations to new API
- Switch mobile autograd to new API

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396128](https://our.internmc.facebook.com/intern/diff/D20396128)

[ghstack-poisoned]

ezyang committed:
Update on "Improved error checking"


- Bring back alias analysis kind check when you def() an operator multiple times.
  It is an error to rely on alias analysis kind merging.  Fixed up op_registration_test
  to respect this.
- Add new syntax sugar dispatch_autograd for registering autograd kernels
- Don't generate schema registration for manual registrations
  (function_wrapper.py)
- Switch VariableType registrations to new API
- Switch mobile autograd to new API

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396128](https://our.internmc.facebook.com/intern/diff/D20396128)

[ghstack-poisoned]

ezyang commented:
all CR comments are addressed (locally in my branch) or responded to

ezyang committed:
Update on "Improved error checking"


- Bring back alias analysis kind check when you def() an operator multiple times.
  It is an error to rely on alias analysis kind merging.  Fixed up op_registration_test
  to respect this.
- Add new syntax sugar dispatch_autograd for registering autograd kernels
- Don't generate schema registration for manual registrations
  (function_wrapper.py)
- Switch VariableType registrations to new API
- Switch mobile autograd to new API

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D20396128](https://our.internmc.facebook.com/intern/diff/D20396128)

[ghstack-poisoned]
